staticaddr/withdraw/manager.go

```diff
@@ -0,0 +1,726 @@
+	outpoints []wire.OutPoint, signer lndclient.SignerClient,
+	musig2Sessions []*input.MuSig2SessionInfo,
+	tx *wire.MsgTx, serverSigs [][]byte) (*wire.MsgTx, error) {
+
+	for idx := range outpoints {
+		haveAllSigs, finalSig, err := signer.MuSig2CombineSig(
+			ctx, musig2Sessions[idx].SessionID,
+			[][]byte{serverSigs[idx]},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if !haveAllSigs {
+			return nil, fmt.Errorf("missing sigs")
+		}
+
+		tx.TxIn[idx].Witness = wire.TxWitness{finalSig}
+	}
+
+	return tx, nil
+}
+
+func toServerOutpoints(outpoints []wire.OutPoint) []*staticaddressrpc.ServerOutPoint { //nolint:lll
+	var result []*staticaddressrpc.ServerOutPoint
+	for _, o := range outpoints {
+		outP := o
+		outpoint := &staticaddressrpc.ServerOutPoint{
+			TxidBytes:   outP.Hash[:],
+			OutputIndex: outP.Index,
+		}
+		result = append(result, outpoint)
+	}
+
+	return result
+}
+
+// createMusig2Sessions creates a musig2 session for a number of deposits.
+func (m *Manager) createMusig2Sessions(ctx context.Context,
+	deposits []*deposit.Deposit) ([]*input.MuSig2SessionInfo, [][]byte,
+	error) {
+
+	musig2Sessions := make([]*input.MuSig2SessionInfo, len(deposits))
+	clientNonces := make([][]byte, len(deposits))
+
+	// Create the sessions and nonces from the deposits.
+	for i := 0; i < len(deposits); i++ {
+		session, err := m.createMusig2Session(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		musig2Sessions[i] = session
+		clientNonces[i] = session.PublicNonce[:]
+	}
+
+	return musig2Sessions, clientNonces, nil
+}
+
+// Musig2CreateSession creates a musig2 session for the deposit.
+func (m *Manager) createMusig2Session(ctx context.Context) (
+	*input.MuSig2SessionInfo, error) {
+
+	addressParams, err := m.cfg.AddressManager.GetStaticAddressParameters(
+		ctx,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get confirmation height for "+
+			"deposit, %v", err)
+	}
+
+	signers := [][]byte{
+		addressParams.ClientPubkey.SerializeCompressed(),
+		addressParams.ServerPubkey.SerializeCompressed(),
+	}
+
+	address, err := m.cfg.AddressManager.GetStaticAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get confirmation height for "+
+			"deposit, %v", err)
+	}
+
+	expiryLeaf := address.TimeoutLeaf
+
+	rootHash := expiryLeaf.TapHash()
+
+	return m.cfg.Signer.MuSig2CreateSession(
+		ctx, input.MuSig2Version100RC2, &addressParams.KeyLocator,
+		signers, lndclient.MuSig2TaprootTweakOpt(rootHash[:], false),
+	)
+}
+
+func (m *Manager) toPrevOuts(deposits []*deposit.Deposit,
+	pkScript []byte) map[wire.OutPoint]*wire.TxOut {
+
+	prevOuts := make(map[wire.OutPoint]*wire.TxOut, len(deposits))
+	for _, d := range deposits {
+		outpoint := wire.OutPoint{
+			Hash:  d.Hash,
+			Index: d.Index,
+		}
+		txOut := &wire.TxOut{
+			Value:    int64(d.Value),
+			PkScript: pkScript,
+		}
+		prevOuts[outpoint] = txOut
+	}
+
+	return prevOuts
+}
+
+func (m *Manager) republishWithdrawals() error {
+	m.Lock()
+	defer m.Unlock()
```

Comment:
m.publishWithdrawal call also happens under the lock. It is a long running call, so I suggest to avoid calling it under the lock.